### ISSUE TITLE
What is the proper way to vendor go-swagger?

### ISSUE THREAD

**Issue by Ravenwater** [go-swagger/go-swagger#730]
I am trying to vendor an API generated by go-swagger to be included in a Docker container. 

> govendor list 
> .....

shows many packages, what is the minimum set?

**Comment by casualjim** [comment]
For this generation to compile you need to have some packages in your GOPATH:

  * github.com/go-openapi/runtime
  * github.com/tylerb/graceful
  * github.com/jessevdk/go-flags
  * golang.org/x/net/context

So that boils down to you have to vendor those packages and you should be good. the go-swagger repo contains all the CLI stuff for the generators, but that's more like build tool not something you'd depend on in code.

**Comment by Ravenwater** [comment]
What about github.com/gorilla/mux, being called from the routers.go? This is where it goes all wrong for me:
root@14d0c9d922ba:/go/src/medicarepathfinder/scoring_engine_mvp# go build api/v0.1.0/swagger-editor-gen/go-server/main.go
api/v0.1.0/swagger-editor-gen/go-server/go/routers.go:6:2: cannot find package "github.com/gorilla/mux" in any of:
        /go/src/medicarepathfinder/scoring_engine_mvp/vendor/github.com/gorilla/mux (vendor tree)
        /usr/local/go/src/github.com/gorilla/mux (from $GOROOT)
        /go/src/github.com/gorilla/mux (from $GOPATH)

I then vendored github.com/gorilla/mux and rebuild the container:
C:\Users\Theodore\dev\Golang\src\github.com\medicarepathfinder\scoring_engine_mvp>govendor list +v
 v  github.com/go-openapi/runtime
 v  github.com/gorilla/mux
 v  github.com/jessevdk/go-flags
 v  github.com/tylerb/graceful
 v  golang.org/x/net/context

root@c64e272829ae:/go/src/medicarepathfinder/scoring_engine_mvp# go build api/v0.1.0/swagger-editor-gen/go-server/main.go
api/v0.1.0/swagger-editor-gen/go-server/go/routers.go:6:2: cannot find package "_/go/src/medicarepathfinder/scoring_engine_mvp/vendor/github.com/gorilla/mux" in any of:
        /usr/local/go/src/_/go/src/medicarepathfinder/scoring_engine_mvp/vendor/github.com/gorilla/mux (from $GOROOT)
        /go/src/_/go/src/medicarepathfinder/scoring_engine_mvp/vendor/github.com/gorilla/mux (from $GOPATH)

For some reason that github.com/gorilla/mux package is not getting referenced properly: the package that is searched for is

-> "_/go/src/medicarepathfinder/scoring_engine_mvp/vendor/github.com/gorilla/mux"
instead of
-> "github.com/gorilla/mux"

I don't know where that "_/go/src/bla" is coming from.

**Comment by Ravenwater** [comment]
I am trying to isolate this problem, so I built a little gorilla/mux server to see if I could reproduce the problems above.  And I can't yet, so at this point, it is still hypothesized to be a problem in the swagger way of doing organizing the code.

C:\Users\Theodore\dev\Golang\src\github.com\Ravenwater\gorilla_mux_test>tree vendor
Folder PATH listing for volume Windows8_OS
Volume serial number is 4EEE-9530
C:\USERS\THEODORE\DEV\GOLANG\SRC\GITHUB.COM\RAVENWATER\GORILLA_MUX_TEST\VENDOR
└───github.com
    └───gorilla
        ├───context
        └───mux

C:\Users\Theodore\dev\Golang\src\github.com\Ravenwater\gorilla_mux_test>govendor list
 v  github.com/gorilla/context
 v  github.com/gorilla/mux
pl  github.com/Ravenwater/gorilla_mux_test
C:\Users\Theodore\dev\Golang\src\github.com\Ravenwater\gorilla_mux_test>go build main.go
C:\Users\Theodore\dev\Golang\src\github.com\Ravenwater\gorilla_mux_test>ls
main.exe  main.go  vendor

**Comment by willis7** [comment]
I use [glide](https://github.com/Masterminds/glide) and a simple `glide init` followed by `glide install` after generating the code worked perfectly. The dependencies were correctly identified and downloaded no problem.

**Comment by l3x** [comment]
I used glide but got the following error:

```
[ERROR]	Error scanning github.com/go-swagger/go-swagger/examples/authentication/client/customers: open /Users/lex/.glide/cache/src/https-github.com-go-swagger-go-swagger/examples/authentication/client/customers: no such file or directory
[ERROR]	This error means the referenced package was not found.
[ERROR]	Missing file or directory errors usually occur when multiple packages
[ERROR]	share a common dependency and the first reference encountered by the scanner
[ERROR]	sets the version to one that does not contain a subpackage needed required
[ERROR]	by another package that uses the shared dependency. Try setting a
[ERROR]	version in your glide.yaml that works for all packages that share this
[ERROR]	dependency.
```

What would be the exact configuration line to use in the glide.yaml file (or other ideas) to fix this?

Thanks, in advance.

**Comment by casualjim** [comment]
I use this yaml in my projects for the go-swagger stuff.

You're not meant to vendor go-swagger directly, there is a much smaller runtime package. In glide I have to list out its dependencies.

```yaml
package: github.com/casualjim/private
import:
- package: github.com/go-openapi/errors
- package: github.com/go-openapi/loads
- package: github.com/go-openapi/runtime
  subpackages:
  - client
  - flagext
  - middleware
- package: github.com/go-openapi/spec
- package: github.com/go-openapi/strfmt
- package: github.com/go-openapi/swag
- package: github.com/go-openapi/validate
- package: github.com/tylerb/graceful
  version: ^1.2.15
- package: golang.org/x/net
  subpackages:
  - context
- package: github.com/mailru/easyjson
  version: master
  subpackages:
  - buffer
  - jlexer
  - jwriter
```

**Comment by l3x** [comment]
@casualjim Thank you for your response!   

Please help me understand the correct/better way to install these go-swagger examples.

I assumed that I would start here: https://github.com/go-swagger/go-swagger

... where I find this URL: https://github.com/go-swagger/go-swagger.git

```
$ cd ~/go/src/bitbucket.org/myrepo
$ git clone https://github.com/go-swagger/go-swagger.git
$ cd go-swagger
$ mkdir src
$ mv {examples,fixtures,generator,cmd,scan} src/
```
Then, I clean up (what I find to be verbose) references:

FROM
`"github.com/go-swagger/go-swagger/examples/todo-list/restapi"`

TO
`"examples/todo-list/restapi"`

Then I set my GOPATH to ~/go:~/go/src/bitbucket.org/myrepo/go-swagger

**Comment by l3x** [comment]
@casualjim Thank you for your response.  

As it turns out, the way go references the vendor directory is not compatible with adding a project path to the GOPATH.

I'd get the [draconian "unexpected directory layout" error](https://github.com/golang/go/issues/17597) when runnning `go install`

```
unexpected directory layout:
	import path: github.com/pkg/errors
	root: /Users/lex/go/src
	dir: /Users/lex/go/src/hi-project/vendor/github.com/pkg/errors
	expand root: /Users/lex/go/src
	expand dir: /Users/lex/go/src/hi-project/vendor/github.com/pkg/errors
	separator: /
```
FWIW - I created an alias that is working well for me:

`alias mvglide='mkdir -p vendors && mv vendor/ vendors/src/ && export GOPATH=$(pwd):$(pwd)/vendors;echo "vendor packages have been moved to $(pwd)/vendors and your GOPATH: $GOPATH"'`

**Comment by casualjim** [comment]
Your comments make me think you're starting out in golang, if that's not the case and you are fully aware of the path you're walking down I applogize.

The best piece of advice I can give you to get going with go is:

**Do not fight the gopath, only misery lies there** should you still want to do that then look at the gb tool

Second piece of advice:

Do not fight gofmt, we all use it nobody likes tabs but it's consistent for everybody that way

**Comment by l3x** [comment]
I can see where you might get that impression.

I did recently start using Glide.

If the good folks that implemented Go's current vendoring implementation understood that the GOPATH should support more than one directory in the path, where each path is colon separated, much like the system PATH, there would be no need for my alias hacks. 

I used to use full github/bitbucket references when referring to imports in my project like this:

 `import "github.com/myrepo/myproject/mysymbol"`

But once I started applying the techniques used in this article... http://manuel.kiessling.net/2012/09/28/applying-the-clean-architecture-to-go-applications/

...I came to appreciate that less is exponentially more:

`import "mysymbol"`

If a directory has a src subdirectory, it should be a candidate for a GOPATH entry.

After playing with glide and vendoring I found that is not the case.
